File: PravahDridh/backend/app/api/v1/endpoints/transactions.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.core.deps import get_current_user, require_roles
from app.db.session import get_db
from app.models.account import Account
from app.models.alert import Alert
from app.models.atm import ATMLocation
from app.models.complaint import Complaint
from app.models.investigation import Investigation
from app.models.prediction import RiskPrediction
from app.models.transaction import SuspiciousTransaction
from app.models.user import User
from app.schemas.common import StandardResponse
from app.schemas.transaction import (
    ActionableIntelligencePackage,
    IngestionStatusResponse,
    SuspiciousTransactionRead,
    TransactionAnalysisResponse,
    TransactionItemRead,
)
from app.services.transaction_analysis_service import TransactionAnalysisService

router = APIRouter()


@router.get("/analysis", response_model=StandardResponse[TransactionAnalysisResponse])
async def get_transaction_analysis(
    alert_id: Optional[uuid.UUID] = Query(None, description="Resolve transactions for this alert"),
    prediction_id: Optional[uuid.UUID] = Query(None, description="Resolve transactions for this prediction"),
    atm_id: Optional[uuid.UUID] = Query(None, description="Target ATM ID"),
    account_id: Optional[uuid.UUID] = Query(None, description="Filter by account"),
    start_time: Optional[datetime] = Query(None, description="Filter from timestamp"),
    end_time: Optional[datetime] = Query(None, description="Filter until timestamp"),
    min_amount: Optional[Decimal] = Query(None, ge=0, description="Minimum amount in INR"),
    max_amount: Optional[Decimal] = Query(None, ge=0, description="Maximum amount in INR"),
    transaction_type: Optional[str] = Query(None, description="Filter type: IMPS, UPI, NEFT, ATM_WITHDRAW"),
    staging_window_hours: Optional[int] = Query(None, ge=1, le=720, description="Configurable pre-window staging buffer (hours)"),
    proximity_radius_meters: Optional[float] = Query(None, ge=50.0, le=100000.0, description="Configurable PostGIS spatial proximity (meters)"),
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=200),
    sort_by: str = Query("occurred_at", description="Sort by occurred_at, amount, velocity_score, relevance_score"),
    sort_order: str = Query("desc", description="Sort order: asc or desc"),
    current_user: User = Depends(require_roles(["INVESTIGATOR", "SUPERVISOR", "ADMIN", "ANALYST"])),
    db: AsyncSession = Depends(get_db),
):
    """
    ALERT → PREDICTION → ATM → WINDOW → RELEVANT TRANSACTIONS → ACCOUNTS → EVIDENCE
    Core Phase 4 investigator analytics endpoint. Computes deterministic relevance scoring,
    delineates prediction basis vs forensic evidence, detects suspicious indicators,
    and returns actionable intelligence packages.
    """
    # Validate sort_by
    allowed_sorts = {"occurred_at", "amount", "velocity_score", "relevance_score"}
    if sort_by not in allowed_sorts:
        sort_by = "occurred_at"

    analysis = await TransactionAnalysisService.analyze_for_alert(
        db=db,
        alert_id=alert_id,
        prediction_id=prediction_id,
        atm_id=atm_id,
        account_id=account_id,
        start_time=start_time,
        end_time=end_time,
        min_amount=min_amount,
        max_amount=max_amount,
        transaction_type=transaction_type,
        staging_window_hours=staging_window_hours,
        proximity_radius_meters=proximity_radius_meters,
        page=page,
        per_page=per_page,
        sort_by=sort_by,
        sort_order=sort_order,
        current_user=current_user,
    )
    return StandardResponse(data=analysis)


@router.get("/analysis/package", response_model=StandardResponse[ActionableIntelligencePackage])
async def get_actionable_intelligence_package(
    alert_id: Optional[uuid.UUID] = Query(None, description="Alert ID for the intelligence package"),
    prediction_id: Optional[uuid.UUID] = Query(None, description="Prediction ID for the intelligence package"),
    current_user: User = Depends(require_roles(["INVESTIGATOR", "SUPERVISOR", "ADMIN", "ANALYST"])),
    db: AsyncSession = Depends(get_db),
):
    """
    Actionable Intelligence Package export endpoint.
    Produces a structured, signed intelligence package for inter-agency relay
    across LEA desks, Bank/FI nodal officers, and I4C coordination units.
    """
    if not alert_id and not prediction_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Either alert_id or prediction_id must be provided to generate the intelligence package",
        )

    analysis = await TransactionAnalysisService.analyze_for_alert(
        db=db,
        alert_id=alert_id,
        prediction_id=prediction_id,
        current_user=current_user,
    )
    if not analysis.actionable_package:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Could not compile actionable intelligence package",
        )
    return StandardResponse(data=analysis.actionable_package)


@router.get("/ingestion/status", response_model=StandardResponse[IngestionStatusResponse])
async def get_ingestion_status(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """
    Get live ingestion pipeline status and database entity counts.
    """
    total_tx = (await db.execute(select(func.count(SuspiciousTransaction.id)))).scalar_one() or 0
    flagged_tx = (await db.execute(select(func.count(SuspiciousTransaction.id)).where(SuspiciousTransaction.is_flagged == True))).scalar_one() or 0
    cash_out_tx = (await db.execute(select(func.count(SuspiciousTransaction.id)).where(SuspiciousTransaction.is_cash_out == True))).scalar_one() or 0

    total_acc = (await db.execute(select(func.count(Account.id)))).scalar_one() or 0
    mule_acc = (await db.execute(select(func.count(Account.id)).where(Account.is_mule_suspected == True))).scalar_one() or 0

    total_comp = (await db.execute(select(func.count(Complaint.id)))).scalar_one() or 0
    total_atm = (await db.execute(select(func.count(ATMLocation.id)))).scalar_one() or 0
    total_pred = (await db.execute(select(func.count(RiskPrediction.id)))).scalar_one() or 0
    total_alert = (await db.execute(select(func.count(Alert.id)))).scalar_one() or 0
    total_inv = (await db.execute(select(func.count(Investigation.id)))).scalar_one() or 0

    earliest_res = await db.execute(select(func.min(SuspiciousTransaction.occurred_at)))
    earliest_tx = earliest_res.scalar_one_or_none()

    latest_res = await db.execute(select(func.max(SuspiciousTransaction.occurred_at)))
    latest_tx = latest_res.scalar_one_or_none()

    status_data = IngestionStatusResponse(
        status="ACTIVE",
        total_transactions=total_tx,
        flagged_transactions=flagged_tx,
        cash_out_transactions=cash_out_tx,
        total_accounts=total_acc,
        mule_suspected_accounts=mule_acc,
        total_complaints=total_comp,
        total_atms=total_atm,
        total_predictions=total_pred,
        total_alerts=total_alert,
        total_investigations=total_inv,
        earliest_transaction=earliest_tx,
        latest_transaction=latest_tx,
        active_dataset="indian_banking_transactions_clean.csv",
    )
    return StandardResponse(data=status_data)


@router.get("/{transaction_id}", response_model=StandardResponse[TransactionItemRead])
async def get_transaction_detail(
    transaction_id: uuid.UUID,
    current_user: User = Depends(require_roles(["INVESTIGATOR", "SUPERVISOR", "ADMIN", "ANALYST"])),
    db: AsyncSession = Depends(get_db),
):
    """
    Retrieve single transaction dossier with eager-loaded relationships and masked account privacy.
    """
    result = await db.execute(
        select(SuspiciousTransaction)
        .options(
            selectinload(SuspiciousTransaction.complaint),
            selectinload(SuspiciousTransaction.account),
            selectinload(SuspiciousTransaction.atm),
        )
        .where(SuspiciousTransaction.id == transaction_id)
    )
    tx = result.scalars().first()
    if not tx:
        raise HTTPException(status_code=404, detail="Transaction not found")

    masked_acc = (
        TransactionAnalysisService.mask_account_hash(tx.account.account_hash)
        if tx.account
        else "ACC-UNKNOWN"
    )

    item = TransactionItemRead(
        id=tx.id,
        complaint_id=tx.complaint_id,
        complaint_number=tx.complaint.complaint_number if tx.complaint else None,
        account_id=tx.account_id,
        source_account_masked=masked_acc,
        bank_name=tx.account.bank_name if tx.account else None,
        account_risk_tier=tx.account.risk_tier if tx.account else None,
        is_mule_suspected=bool(tx.account.is_mule_suspected if tx.account else False),
        amount=tx.amount,
        transaction_type=tx.transaction_type,
        occurred_at=tx.occurred_at,
        latitude=tx.latitude,
        longitude=tx.longitude,
        atm_id=tx.atm_id,
        destination_atm_code=tx.atm.atm_code if tx.atm else None,
        distance_to_atm_meters=None,
        velocity_score=tx.velocity_score or Decimal("0.0"),
        is_flagged=tx.is_flagged,
        is_cash_out=tx.is_cash_out,
        relevance="HIGH" if tx.is_cash_out else "MEDIUM",
        relevance_score=0.85 if tx.is_cash_out else 0.50,
        relevance_reasons=["Direct dossier inspection"],
        created_at=tx.created_at or tx.occurred_at,
    )
    return StandardResponse(data=item)

File: PravahDridh/frontend/src/api/transactions.ts
/**
 * PravahDridh — Real Transaction Intelligence & Forensic Evidence API Service
 * Interacts with /api/v1/transactions/*
 */

import { apiClient, unwrapResponse } from './client';
import { StandardResponse } from '../types/common';

export interface SuspiciousIndicator {
  indicator: string;
  label: string;
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  observed_facts: Record<string, any>;
  suspicious_interpretation: string;
  evidence: Record<string, any>;
}

export interface InvestigativeRecommendation {
  action_id: string;
  priority: 'URGENT' | 'HIGH' | 'MEDIUM' | 'ROUTINE';
  target_entity: string;
  title: string;
  description: string;
  rationale: string;
}

export interface InvestigationContext {
  alert_id?: string;
  alert_severity?: string;
  alert_status?: string;
  prediction_id?: string;
  model_version?: string;
  atm_id?: string;
  atm_code?: string;
  bank_name?: string;
  city?: string;
  latitude?: number;
  longitude?: number;
  risk_score?: number;
  confidence?: number;
  predicted_window_start?: string;
  predicted_window_end?: string;
  prediction_basis: string[];
}

export interface MaskedAccountSummary {
  account_id: string;
  account_masked: string;
  bank_name: string;
  account_type: string;
  risk_tier: string;
  is_mule_suspected: boolean;
  transaction_count: number;
  total_volume: number;
}

export interface TransactionItem {
  id: string;
  complaint_id?: string;
  complaint_number?: string;
  account_id?: string;
  source_account_masked?: string;
  bank_name?: string;
  account_risk_tier?: string;
  is_mule_suspected: boolean;
  amount: number;
  transaction_type: string;
  occurred_at: string;
  latitude: number;
  longitude: number;
  atm_id?: string;
  destination_atm_code?: string;
  distance_to_atm_meters?: number;
  velocity_score: number;
  is_flagged: boolean;
  is_cash_out: boolean;
  relevance: 'HIGH' | 'MEDIUM' | 'LOW';
  relevance_score: number;
  relevance_reasons: string[];
  created_at: string;
}

export interface TransactionSummary {
  total_transactions: number;
  total_amount: number;
  unique_accounts: number;
  unique_destinations: number;
  cash_out_count: number;
  earliest_transaction?: string;
  latest_transaction?: string;
  atm_concentration_pct: number;
}

export interface ActionableIntelligencePackage {
  package_id: string;
  generated_at: string;
  alert_id: string;
  prediction_id: string;
  severity: string;
  risk_score: number;
  confidence: number;
  predicted_atm: {
    id?: string;
    atm_code?: string;
    bank_name?: string;
    city?: string;
    latitude?: number;
    longitude?: number;
  };
  prediction_window: {
    start?: string;
    end?: string;
  };
  prediction_basis: string[];
  forensic_summary: TransactionSummary;
  relevant_accounts: MaskedAccountSummary[];
  verified_indicators: SuspiciousIndicator[];
  recommended_actions: InvestigativeRecommendation[];
  dissemination_targets: string[];
}

export interface TransactionAnalysisResponse {
  context: InvestigationContext;
  summary: TransactionSummary;
  transactions: TransactionItem[];
  indicators: SuspiciousIndicator[];
  recommendations: InvestigativeRecommendation[];
  relevant_accounts: MaskedAccountSummary[];
  actionable_package?: ActionableIntelligencePackage;
  page: number;
  per_page: number;
  total: number;
}

export interface TransactionAnalysisParams {
  alert_id?: string;
  prediction_id?: string;
  atm_id?: string;
  account_id?: string;
  start_time?: string;
  end_time?: string;
  min_amount?: number;
  max_amount?: number;
  transaction_type?: string;
  staging_window_hours?: number;
  proximity_radius_meters?: number;
  page?: number;
  per_page?: number;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export const transactionsApi = {
  /**
   * Run real alert-driven transaction analysis and evidence retrieval
   * GET /api/v1/transactions/analysis
   */
  async analyze(params: TransactionAnalysisParams = {}): Promise<TransactionAnalysisResponse> {
    const q = new URLSearchParams();
    if (params.alert_id) q.append('alert_id', params.alert_id);
    if (params.prediction_id) q.append('prediction_id', params.prediction_id);
    if (params.atm_id) q.append('atm_id', params.atm_id);
    if (params.account_id) q.append('account_id', params.account_id);
    if (params.start_time) q.append('start_time', params.start_time);
    if (params.end_time) q.append('end_time', params.end_time);
    if (params.min_amount !== undefined) q.append('min_amount', String(params.min_amount));
    if (params.max_amount !== undefined) q.append('max_amount', String(params.max_amount));
    if (params.transaction_type && params.transaction_type !== 'ALL') {
      q.append('transaction_type', params.transaction_type);
    }
    if (params.staging_window_hours !== undefined) {
      q.append('staging_window_hours', String(params.staging_window_hours));
    }
    if (params.proximity_radius_meters !== undefined) {
      q.append('proximity_radius_meters', String(params.proximity_radius_meters));
    }
    if (params.page) q.append('page', String(params.page));
    if (params.per_page) q.append('per_page', String(params.per_page));
    if (params.sort_by) q.append('sort_by', params.sort_by);
    if (params.sort_order) q.append('sort_order', params.sort_order);

    const queryStr = q.toString() ? `?${q.toString()}` : '';
    return unwrapResponse<TransactionAnalysisResponse>(
      apiClient.get<StandardResponse<TransactionAnalysisResponse>>(`/transactions/analysis${queryStr}`)
    );
  },

  /**
   * Fetch standalone Actionable Intelligence Package
   * GET /api/v1/transactions/analysis/package
   */
  async getIntelligencePackage(
    alertId?: string,
    predictionId?: string
  ): Promise<ActionableIntelligencePackage> {
    const q = new URLSearchParams();
    if (alertId) q.append('alert_id', alertId);
    if (predictionId) q.append('prediction_id', predictionId);

    return unwrapResponse<ActionableIntelligencePackage>(
      apiClient.get<StandardResponse<ActionableIntelligencePackage>>(
        `/transactions/analysis/package?${q.toString()}`
      )
    );
  },

  /**
   * Fetch single transaction detail dossier
   * GET /api/v1/transactions/{transaction_id}
   */
  async getDetail(transactionId: string): Promise<TransactionItem> {
    return unwrapResponse<TransactionItem>(
      apiClient.get<StandardResponse<TransactionItem>>(`/transactions/${transactionId}`)
    );
  },

  /**
   * Fetch live ingestion pipeline status and database entity counts
   * GET /api/v1/transactions/ingestion/status
   */
  async getIngestionStatus(): Promise<IngestionStatus> {
    return unwrapResponse<IngestionStatus>(
      apiClient.get<StandardResponse<IngestionStatus>>('/transactions/ingestion/status')
    );
  },
};

export interface IngestionStatus {
  status: string;
  total_transactions: number;
  flagged_transactions: number;
  cash_out_transactions: number;
  total_accounts: number;
  mule_suspected_accounts: number;
  total_complaints: number;
  total_atms: number;
  total_predictions: number;
  total_alerts: number;
  total_investigations: number;
  earliest_transaction?: string | null;
  latest_transaction?: string | null;
  active_dataset: string;
}

